Use transport actions instead of guice for xpack info

rjernst opened the pull request:
The Xpack info API internally gathers all xpack feature sets by using
guice to bind many implementations. The info api then iterates over
these feature sets to gather usage information for each. However, since
each of these implementations are in different plugins, there is no
direct, non-guice way for them to register the implementations. This
commit removes the dependence on guice for the info api by having the
core xpack jar contain an action instance for each plugin, and the
implementation of those actions are then registered within each xpack
feature plugin. The info api then iterates over each of these actions
and calls them with the NodeClient.

rjernst committed:
Use transport actions instead of guice for xpack info

The Xpack info API internally gathers all xpack feature sets by using
guice to bind many implementations. The info api then iterates over
these feature sets to gather usage information for each. However, since
each of these implementations are in different plugins, there is no
direct, non-guice way for them to register the implementations. This
commit removes the dependence on guice for the info api by having the
core xpack jar contain an action instance for each plugin, and the
implementation of those actions are then registered within each xpack
feature plugin. The info api then iterates over each of these actions
and calls them with the NodeClient.

rjernst committed:
fix javadoc

rjernst committed:
fix dataframe info action registration

rjernst committed:
Merge branch 'master' into deguice47

talevy commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/XPackInfoFeatureTransportAction.java:
```diff
@@ -0,0 +1,32 @@
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+
+public abstract class XPackInfoFeatureTransportAction extends HandledTransportAction<XPackInfoRequest, XPackInfoFeatureResponse> {
```
should we add javadocs to this class since I suspect it will be picked up by new devs creating new plugins that aren't familiar with it

talevy reviewed (commented):
overall looks good! just a nit and request for more javadoc

I suspect there are existing xpack info rest tests that are unchanged, right?

rjernst committed:
rework subactions iteration

rjernst committed:
fix unit test

rjernst committed:
add javadoc

rjernst committed:
fix accidental newline

rjernst commented:
@talevy I addressed your comments. In answer to your question about tests: yes, there are existing tests for all the info action implementations. These are just the feature set tests we had before but renamed.

talevy reviewed (approved):
LGTM pending some test fixes

rjernst committed:
don't pass null for license state

rjernst committed:
fix watcher assertion

rjernst committed:
fix null actionfilters

rjernst committed:
fix security test

rjernst committed:
fix sql test

elasticmachine commented:
Pinging @elastic/es-core-infra
